plugin/executor/containerd/containerd.go

```diff
@@ -25,153 +23,141 @@ type ExitHandler interface {
-	e := &Executor{
-		rootDir:     rootDir,
+func New(ctx context.Context, stateDir string, cli *containerd.Client, exitHandler ExitHandler) (*Executor, error) {
+	return &Executor{
+		stateDir:    stateDir,
 		exitHandler: exitHandler,
-	}
-
-	client, err := libcontainerd.NewClient(ctx, cli, rootDir, PluginNamespace, e)
-	if err != nil {
-		return nil, errors.Wrap(err, "error creating containerd exec client")
-	}
-	e.client = client
-	return e, nil
+		client:      cli,
+	}, nil
 }
 
 // Executor is the containerd client implementation of a plugin executor
 type Executor struct {
-	rootDir     string
-	client      Client
+	stateDir    string
+	client      *containerd.Client
 	exitHandler ExitHandler
 }
 
-// deleteTaskAndContainer deletes plugin task and then plugin container from containerd
-func deleteTaskAndContainer(ctx context.Context, cli Client, id string) {
-	_, _, err := cli.DeleteTask(ctx, id)
-	if err != nil && !errdefs.IsNotFound(err) {
-		logrus.WithError(err).WithField("id", id).Error("failed to delete plugin task from containerd")
-	}
-
-	err = cli.Delete(ctx, id)
-	if err != nil && !errdefs.IsNotFound(err) {
-		logrus.WithError(err).WithField("id", id).Error("failed to delete plugin container from containerd")
-	}
-}
-
 // Create creates a new container
 func (e *Executor) Create(id string, spec specs.Spec, stdout, stderr io.WriteCloser) error {
-	opts := runctypes.RuncOptions{
-		RuntimeRoot: filepath.Join(e.rootDir, "runtime-root"),
+	opts := &runctypes.RuncOptions{
+		RuntimeRoot: filepath.Join(e.stateDir, "runtime-root"),
 	}
 	ctx := context.Background()
-	err := e.client.Create(ctx, id, &spec, &opts)
+	container, err := e.client.NewContainer(ctx, id,
+		containerd.WithSpec(&spec),
+		containerd.WithRuntime(fmt.Sprintf("io.containerd.runtime.v1.%s", runtime.GOOS), opts),
+	)
 	if err != nil {
-		status, err2 := e.client.Status(ctx, id)
-		if err2 != nil {
-			if !errdefs.IsNotFound(err2) {
-				logrus.WithError(err2).WithField("id", id).Warn("Received an error while attempting to read plugin status")
-			}
-		} else {
-			if status != libcontainerd.StatusRunning && status != libcontainerd.StatusUnknown {
-				if err2 := e.client.Delete(ctx, id); err2 != nil && !errdefs.IsNotFound(err2) {
-					logrus.WithError(err2).WithField("plugin", id).Error("Error cleaning up containerd container")
-				}
-				err = e.client.Create(ctx, id, &spec, &opts)
-			}
-		}
-
-		if err != nil {
-			return errors.Wrap(err, "error creating containerd container")
-		}
+		return err
 	}
-
-	_, err = e.client.Start(ctx, id, "", false, attachStreamsFunc(stdout, stderr))
+	task, err := container.NewTask(ctx, cio.NewCreator(cio.WithStreams(nil, stdout, stderr)))
 	if err != nil {
-		deleteTaskAndContainer(ctx, e.client, id)
+		container.Delete(ctx)
+		return err
 	}
-	return err
+	wait, err := task.Wait(ctx)
+	if err != nil {
+		task.Delete(ctx)
+		container.Delete(ctx)
```

Comment:
Errors returned by `task.Delete()` and `container.Delete()` don't have to be handled? (Here, and other cases below).